main.go

```diff
@@ -3,126 +3,158 @@ package main
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	go func() {
-		defer cancel()
+	defer cancel()
 
-		for {
-			tcpConn, err := listener.Accept()
-			if err != nil {
-				fmt.Printf("Error accepting: %s\n", err.Error())
+	for {
+		tcpConn, err := listener.Accept()
+		if err != nil {
+			logrus.Errorf("Error accepting: %s\n", err.Error())
 
-				return
-			}
+			return
+		}
 
-			fmt.Printf("Received message from %s to %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
+		logrus.Debugf("(Connection from %s to %s) Starting\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
 
-			wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
-				"Authorization": []string{
-					"Bearer " + apiKey,
-				},
-				"X-Application-ID": []string{applicationID},
-			})
+		wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
+			"Authorization": []string{
+				"Bearer " + apiKey,
+			},
+			"X-Application-ID": []string{applicationID},
+		})
 
-			if err != nil {
-				fmt.Printf("Error dial into edge tunneling service: %s\n, cloudURI - %s\n", err.Error(), cloudURI)
+		if err != nil {
+			logrus.Errorf("(Connection from %s to %s) Error dial into edge tunneling service: %s\n, cloudURI - %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error(), cloudURI)
 
-				return
-			}
+			return
+		}
 
-			ctx, cancel := context.WithCancel(ctx)
+		_, cancel := context.WithCancel(ctx)
 
-			go func() {
-				for {
+		go func() {
+			chanBytes := make(chan []byte, 4096)
+			go func () {
+				for{
 					bytes := make([]byte, 4096)
 					length, err := tcpConn.Read(bytes)
 					if err != nil {
-						fmt.Printf("failed to read bytes from tcp connection: %s\n", err.Error())
+						logrus.Errorf("(Connection from %s to %s) failed to read bytes from tcp connection: %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error())
+						cancel()
 
-						break
+						return
 					}
+					chanBytes <- bytes[:length]
+				}
+			}()
 
+			c := time.Tick(time.Duration(pingDuration) * time.Second)
+			logrus.Infof("pingDuration :%v c:%#v\n", pingDuration, c)
+
+			for {
+				select {
+				case bytes := <- chanBytes:
+					length := len(bytes)
 					if err := wsConn.WriteMessage(websocket.BinaryMessage, bytes[:length]); err != nil {
-						fmt.Printf("failed to write binary message to websocket connection: %s\n", err.Error())
+							logrus.Errorf("(Connection from %s to %s) failed to write binary message to websocket connection: %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error())
+						cancel()
 
-						break
+						return
 					}
 
-					fmt.Printf("write %d bytes of data to websocket connection\n", length)
-				}
-
-				cancel()
-			}()
+					logrus.Debugf("write %d bytes of data to websocket connection\n", length)
 
-			go func() {
-				for {
-					_, data, err := wsConn.ReadMessage()
-					if err != nil {
-						fmt.Printf("failed to read binary message from websocket connection: %s\n", err.Error())
+				case <-c:
```

Comment:
we might not need to have select statement to handle this. Maybe we should keep whatever we have before and then have one go routine to just send ping messages.

Reply:
I'd rather not send ping if I don't need to send it, so this explains the select here, also added a ctx.Done in the select case.